src/lib/statetable.rs

```diff
@@ -148,6 +101,54 @@ impl StateTable {
         StateTable { actions: actions, gotos: gotos, reduce_reduce: reduce_reduce,
                      shift_reduce: shift_reduce }
     }
+
+}
+
+fn resolve_shift_reduce(grm: &Grammar, mut e: OccupiedEntry<(usize, Symbol), Action>, term_k: TIdx,
+                        prod_k: PIdx, state_j: usize) -> u64 {
+    let mut shift_reduce = 0;
+    let term_k_prec = grm.terminal_precs[term_k];
+    let prod_k_prec = grm.prod_precs[prod_k];
+    match (term_k_prec, prod_k_prec) {
+        (None, None) | (Some(_), None) | (None, Some(_)) => {
+            // If the terminal and production don't both have precedences, we use Yacc's default
+            // resolution, which is in favour of the shift.
+            e.insert(Action::Shift(state_j));
+            shift_reduce += 1;
+        },
+        (Some(Precedence{level: term_lev, kind: term_kind}),
+         Some(Precedence{level: prod_lev, kind: prod_kind})) => {
```

Comment:
This could be simplified to `(Some(term@Precedence{..}), Some(prod@Precedence{..}))`
In the body `term.level`, etc are then available.